mozfldp/fed_avg_w_dp.py

```diff
@@ -0,0 +1,250 @@
+    """
+    Merge all user updates for a round into a single delta (vector).
+    """
+
+    num_users_in_batch = len(user_updates_buf)
+    merged_weights = np.zeros(len(user_updates_buf[0]), dtype=np.float64, order="C")
+
+    for i in range(num_users_in_batch):
+        weighted_user_val = np.multiply(user_updates_buf[i], user_weights[i])
+        merged_weights = np.add(merged_weights, weighted_user_val)
+
+    merged_weights = np.divide(merged_weights, user_sel_prob * weight_sum)
+    return merged_weights
+
+
+def flat_clip(sensitivity, vec):
+    vec *= min(1, sensitivity / np.linalg.norm(vec))
+    return vec
+
+
+def user_update_fed_avg(prms, round_user_features, round_user_labels, theta_0):
+
+    batches_features = []
+    batches_labels = []
+
+    classifier = SGDClassifier(
+        loss="hinge", penalty="l2", max_iter=1, random_state=prms.rand_seed
+    )
+
+    round_num_entries = len(round_user_features)
+
+    # For now just skip any batches that are smaller than batch_size
+    num_batches = round_num_entries // prms.batch_size
+
+    theta = np.array(theta_0, copy=True)
+
+    # Split the round data into seperate batches
+    for i in range(0, round_num_entries, prms.batch_size):
+        batches_features.append(round_user_features[i : i + prms.batch_size])
+        batches_labels.append(round_user_labels[i : i + prms.batch_size])
+
+    for _ in range(prms.num_epochs):
+        for j in range(num_batches):
+
+            coef, inter = _get_coef_and_inter_slice_from_theta(
+                theta, prms.num_features, prms.num_labels
+            )
+
+            classifier.fit(
+                batches_features[j],
+                batches_labels[j],
+                coef_init=coef,
+                intercept_init=inter,
+            )
+
+            _set_coef_and_inter_on_theta(theta, classifier.coef_, classifier.intercept_)
+            theta = theta_0 + flat_clip(prms.sensitivity, theta - theta_0)
+
+    return theta - theta_0
+
+
+def _get_random_selection_of_user_idxs(num_users):
+    # For now just randomly pick 1/5th of the users.
+    # TODO: Change this later!
+    return np.random.choice(num_users, num_users // 5)
+
+
+def _init_theta_from_moment_accountant(num_features, num_labels):
+    # Just a placeholder value for now
+
+    # theta_0 is internally a linear array. We will use functon to get slices of the features/labels.
+    return np.zeros(num_labels * num_features + num_labels, dtype=np.float64, order="C")
+
+
+def _init_user_weights_and_weight_sum(num_users, weight_mod):
+    weights = []
+    init_weight = min(num_users / weight_mod, 1)
+
+    for _ in range(num_users):
+        weights.append(init_weight)
+
+    weight_sum = num_users * init_weight
+
+    return weights, weight_sum
+
+
+def _calc_standard_dev(noise_scale, sensitivity, usr_sel_prob, weight_sum):
+    return (noise_scale * sensitivity) / (usr_sel_prob * weight_sum)
+
+
+def _get_data_for_user_for_round(prms, data, user_id):
```

Comment:
It looks like this limits the data used for a given user in the current round to a random subset of size at least as large as the batch size. This is not necessary - each round should cover the full user data, split into batches (as is done in `user_update_fed_avg`).

Reply:
Done.